lib/lightning/kafka_triggers/message_server.ex

```diff
@@ -0,0 +1,40 @@
+defmodule Lightning.KafkaTriggers.MessageServer do
+  @moduledoc """
+  Server responsible for maintaining a list of messages with nil keys that
+  are provided to the worker processes.
+  """
+  use GenServer
+
+  alias Lightning.KafkaTriggers.MessageHandling
+
+  def start_link(_opts) do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  def next_message do
+    GenServer.call(__MODULE__, :next_message)
+  end
+
+  @impl true
+  def init(_opts) do
+    {:ok, []}
+  end
+
+  @impl true
+  def handle_call(:next_message, _from, current_messages) do
+    {message, remaining_messages} = pop_message(current_messages)
+
+    {:reply, message, remaining_messages}
+  end
+
+  defp pop_message([]) do
+    case MessageHandling.find_nil_key_message_ids() do
```

Comment:
Can we have this function defined in this module? It is only used here and makes the code much easier to review. Also I appreciate your concern with making functions more testable but here even if it's private one can test it by calling `MessageServer.next_message()` or even directly 
`MessageServer.handle_call(:next_message, pid(), [])` if the module state is used for other test.

Reply:
@jyeshe Apologies for making review harder :(. 

I kept it separate to make the tests easier to follow. At the moment, given the 'refresh' mechanism contained within the server I would need to cycle through the messages, have it refresh and then see that we have definitely not gotten a message with a nil key. 

I could have it as a public method on the server and retain the simplicity of testing that it currently enjoys on MessageHandling, but the non-nil key counterpart currently lives on MessageHandling, so it felt more coherent to keep the method there.

At the moment, I am not sure if this code will survive load-testing :). Even if it does, the plan to move to Cachex would mean that much of this code will be changed anyway. So, with all those caveats, could we keep it here for now, please? 

I do not see this as the final implementation of nil-key handling, but rather a temporary implementation to deal with the fact that the load-testing data is sans key. Once I get a gap I would like to implement a better-engineered solution so this PR is just about adding enough wallpaper over the crack to allow us to continue with the load testing.

Reply:
Okay @rorymckinley I appreciate the justification. It's not an issue after reviewing it again and refreshing the pipeline on our mind.